File: MyApplication/app/src/main/java/com/homework/myapplication/controller/MainActivity.java
package com.homework.myapplication.controller;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.homework.myapplication.R;
import com.homework.myapplication.app.MyApplication;
import com.homework.myapplication.bean.Answer;
import com.homework.myapplication.bean.Question;
import com.homework.myapplication.model.DaoManager;
import com.homework.myapplication.util.DataUtil;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class MainActivity extends AppCompatActivity {
    private EditText tvName;
    private TextView tvQuestion;
    private Button btStart;
    private DaoManager daoManager;
    private List<Question> questionList;
    private List<Question> questions = new ArrayList<>();
    private List<Answer> answersList;
    private Button btQuery;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        daoManager = MyApplication.getCurrent().getDaoManager();
        initView();
        initData();
        btStart.setOnClickListener(v->{
            if (TextUtils.isEmpty(tvName.getText())){
                Toast.makeText(this,"Please enter your name",Toast.LENGTH_SHORT).show();
                return;
            }
            questions.clear();
            List<String> themeList = questionList.stream().map(q -> q.getTheme()).distinct().collect(Collectors.toList());
            for (String theme : themeList) {
                List<Question> themeQuestionList = questionList.stream().filter(q -> q.getTheme().trim().equals(theme)).collect(Collectors.toList());
                Collections.shuffle(themeQuestionList);
                List<Question> questionList2 = new ArrayList<>(themeQuestionList.subList(0, 2));
                questions.addAll(questionList2);
            }
            for (Question question : questions) {
                List<Answer> answers = new ArrayList<>();
                for (Answer answer : answersList) {
                    if (question.getId()==answer.getQuestionId()){
                        answers.add(answer);
                    }
                }
                Optional<Answer> first = answers.stream().filter(a -> a.getIsCorrect() == 1).findFirst();
                if (first!=null&&first.isPresent()){
                    Answer correctAnswer = first.get();
                    List<Answer> answerList = answers.stream().filter(a -> a.getIsCorrect() == 0).collect(Collectors.toList());
                    List<Answer> subList = new ArrayList<>(answerList.subList(0,4));
                    subList.add(correctAnswer);
                    Collections.shuffle(subList);
                    question.setAnswerList(subList);
                }
            }

            Intent intent = new Intent(this, ExamActivity.class);
            intent.putExtra("userName",tvName.getText().toString().trim());
            intent.putExtra("questions", (Serializable) questions);
            startActivity(intent);
            for (Question question : questions) {
                for (Answer answer : question.getAnswerList()) {
                    Log.e("question",question.getId()+"---"+question.getTheme()+"----"+question.getQuestionSrc()+"----"+answer.getAnswerContent()+"----"+answer.getIsCorrect());
                }
            }
        });
        btQuery.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this, ResultActivity.class);
                startActivity(intent);
            }
        });
    }

    private void initData() {
        DataUtil.getData(getBaseContext(),daoManager);
        questionList = daoManager.queryQuestionData();
        answersList = daoManager.queryAnswerData();
    }


    private void initView() {
        tvName = (EditText) findViewById(R.id.tv_name);
        btStart = (Button) findViewById(R.id.bt_start);
        btQuery = (Button) findViewById(R.id.bt_query);
        tvQuestion = (TextView) findViewById(R.id.tv_question);
    }
}

File: MyApplication/app/src/main/java/com/homework/myapplication/controller/MainActivity.kt
package com.homework.myapplication.controller

import android.app.Application
import com.homework.myapplication.model.DaoManager
import com.homework.myapplication.interfaces.CallBack
import com.homework.myapplication.bean.Score
import android.app.Activity
import android.content.Intent
import com.homework.myapplication.controller.ResultActivity
import com.homework.myapplication.app.MyApplication
import com.homework.myapplication.bean.Exam
import com.homework.myapplication.bean.Answer
import com.homework.myapplication.util.SharePerferenceUtils
import com.homework.myapplication.util.DataUtil
import com.homework.myapplication.bean.Question
import com.homework.myapplication.util.ListUtil
import android.database.sqlite.SQLiteOpenHelper
import com.homework.myapplication.util.DataBaseHelper
import android.database.sqlite.SQLiteDatabase
import android.content.SharedPreferences
import android.content.ContentValues
import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import android.view.LayoutInflater
import com.homework.myapplication.R
import android.text.TextUtils
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.widget.TextView
import com.homework.myapplication.adapter.ResultAdapter.ResultHolder
import androidx.recyclerview.widget.LinearLayoutManager
import com.homework.myapplication.adapter.QuestionAdapter
import androidx.fragment.app.FragmentPagerAdapter
import android.os.Bundle
import kotlin.jvm.JvmOverloads
import androidx.appcompat.app.AppCompatActivity
import com.homework.myapplication.controller.NoScrollViewPager
import com.homework.myapplication.util.GuidGeneratorUtil
import com.homework.myapplication.controller.ExamFragment
import com.homework.myapplication.adapter.MyFragmentPagerAdapter
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import android.widget.Toast
import com.homework.myapplication.util.PecentUtil
import android.os.Build
import android.util.Log
import com.homework.myapplication.adapter.OptionAdapter
import android.widget.EditText
import com.homework.myapplication.controller.ExamActivity
import com.homework.myapplication.adapter.ResultAdapter
import androidx.viewpager.widget.ViewPager
import android.view.MotionEvent
import android.view.View
import android.widget.Button
import java.io.Serializable
import java.util.*
import java.util.stream.Collectors

class MainActivity : AppCompatActivity() {
    private var tvName: EditText? = null
    private var tvQuestion: TextView? = null
    private var btStart: Button? = null
    private var daoManager: DaoManager? = null
    private var questionList: List<Question?>? = null
    private val questions: MutableList<Question?> = ArrayList()
    private var answersList: List<Answer?>? = null
    private var btQuery: Button? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        daoManager = MyApplication.Companion.getCurrent().getDaoManager()
        initView()
        initData()
        btStart!!.setOnClickListener { v: View? ->
            if (TextUtils.isEmpty(tvName!!.text)) {
                Toast.makeText(this, "Please enter your name", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }
            questions.clear()
            val themeList = questionList!!.stream().map { q: Question? -> q.getTheme() }
                .distinct().collect(Collectors.toList())
            for (theme in themeList) {
                val themeQuestionList = questionList!!.stream()
                    .filter { q: Question? -> q.getTheme().trim { it <= ' ' } == theme }
                    .collect(Collectors.toList())
                Collections.shuffle(themeQuestionList)
                val questionList2: List<Question?> = ArrayList(themeQuestionList.subList(0, 2))
                questions.addAll(questionList2)
            }
            for (question in questions) {
                val answers: MutableList<Answer?> = ArrayList()
                for (answer in answersList!!) {
                    if (question.getId() == answer.getQuestionId()) {
                        answers.add(answer)
                    }
                }
                val first =
                    answers.stream().filter { a: Answer? -> a.getIsCorrect() == 1 }.findFirst()
                if (first != null && first.isPresent) {
                    val correctAnswer = first.get()
                    val answerList =
                        answers.stream().filter { a: Answer? -> a.getIsCorrect() == 0 }.collect(
                            Collectors.toList()
                        )
                    val subList: MutableList<Answer?> = ArrayList(answerList.subList(0, 4))
                    subList.add(correctAnswer)
                    Collections.shuffle(subList)
                    question.setAnswerList(subList)
                }
            }
            val intent = Intent(this, ExamActivity::class.java)
            intent.putExtra("userName", tvName!!.text.toString().trim { it <= ' ' })
            intent.putExtra("questions", questions as Serializable)
            startActivity(intent)
            for (question in questions) {
                for (answer in question.getAnswerList()) {
                    Log.e("question",
                        question.getId()
                            .toString() + "---" + question.getTheme() + "----" + question.getQuestionSrc() + "----" + answer.answerContent + "----" + answer.isCorrect
                    )
                }
            }
        }
        btQuery!!.setOnClickListener {
            val intent = Intent(this@MainActivity, ResultActivity::class.java)
            startActivity(intent)
        }
    }

    private fun initData() {
        DataUtil.getData(baseContext, daoManager)
        questionList = daoManager!!.queryQuestionData()
        answersList = daoManager!!.queryAnswerData()
    }

    private fun initView() {
        tvName = findViewById<View>(R.id.tv_name) as EditText
        btStart = findViewById<View>(R.id.bt_start) as Button
        btQuery = findViewById<View>(R.id.bt_query) as Button
        tvQuestion = findViewById<View>(R.id.tv_question) as TextView
    }
}